# NLP Project

## Project Goals :
- clean the text data [ capiltalization and lower case / remove any urls or special charachters /
remove pancituation ]
- remove stop words 
- toknization 
- Lemmatization
- BERT for classification

### Cleaning data is crucial for NLP as it enhances model accuracy by removing noise, inconsistencies, and irrelevant information, enabling algorithms to focus on meaningful patterns and relationships within the text.

## Importing Libararies Needed

In [ ]:
# for data reading and slicing
import numpy as np
import pandas as pd

# for string manipulate
import string
import re

# Import Natural language Toolkit for stopwords and tokization 
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

# for handling contractions
import contractions

# for train and test split 
from sklearn.model_selection import train_test_split

# Import Natural language Toolkit for lemmatization 
from nltk.stem import WordNetLemmatizer

# for randomly selecting the data
import random

# for BERT Neural Networks inside it
import tensorflow as tf
from keras.models import Model
from keras.layers import Input, Dense
from transformers import BertTokenizer, TFBertModel , TFBertForSequenceClassification
from keras.utils import to_categorical

# for evaluating the model 
from sklearn.metrics import classification_report

c:\Users\User\AppData\Local\Programs\Python\Python311\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [ ]:
# Download NLTK tokenizer data
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [ ]:
# Download NLTK Stop Words data
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

## Reading and Analyzing and Understanding the data

In [ ]:
# Read CSV 
df = pd.read_csv(r"C:\Users\User\Desktop\koleya\summer 2\SPRINTS  AI & ML program\Assigment\assigment 6\train.csv")

# look inside the first 10 rows of data
df.head(50)

id                                       comment_text  \
0   0000997932d777bf  Explanation\nWhy the edits made under my usern...   
1   000103f0d9cfb60f  D'aww! He matches this background colour I'm s...   
2   000113f07ec002fd  Hey man, I'm really not trying to edit war. It...   
3   0001b41b1c6bb37e  "\nMore\nI can't make any real suggestions on ...   
4   0001d958c54c6e35  You, sir, are my hero. Any chance you remember...   
5   00025465d4725e87  "\n\nCongratulations from me as well, use the ...   
6   0002bcb3da6cb337       COCKSUCKER BEFORE YOU PISS AROUND ON MY WORK   
7   00031b1e95af7921  Your vandalism to the Matt Shirvington article...   
8   00037261f536c51d  Sorry if the word 'nonsense' was offensive to ...   
9   00040093b2687caa  alignment on this subject and which are contra...   
10  0005300084f90edc  "\nFair use rationale for Image:Wonju.jpg\n\nT...   
11  00054a5e18b50dd4  bbq \n\nbe a man and lets discuss it-maybe ove...   
12  0005c987bdfc9d4b  Hey... what is it..\n@ | talk .\nWhat is it......   
13  0006f16e4e9f292e  Before you start throwing accusations and warn...   
14  00070ef96486d6f9  Oh, and the girl above started her arguments w...   
15  00078f8ce7eb276d  "\n\nJuelz Santanas Age\n\nIn 2002, Juelz Sant...   
16  0007e25b2121310b  Bye! \n\nDon't look, come or think of comming ...   
17  000897889268bc93   REDIRECT Talk:Voydan Pop Georgiev- Chernodrinski   
18  0009801bd85e5806  The Mitsurugi point made no sense - why not ar...   
19  0009eaea3325de8c  Don't mean to bother you \n\nI see that you're...   
20  000b08c464718505  "\n\n Regarding your recent edits \n\nOnce aga...   
21  000bfd0867774845  "\nGood to know. About me, yeah, I'm studying ...   
22  000c0dfd995809fa  "\n\n Snowflakes are NOT always symmetrical! \...   
23  000c6a3f0cd3ba8e  "\n\n The Signpost: 24 September 2012 \n\n Rea...   
24  000cfee90f50d471  "\n\nRe-considering 1st paragraph edit?\nI don...   
25  000eefc67a2c930f  Radial symmetry \n\nSeveral now extinct lineag...   
26  000f35deef84dc4a  There's no need to apologize. A Wikipedia arti...   
27  000ffab30195c5e1  Yes, because the mother of the child in the ca...   
28  0010307a3a50a353  "\nOk. But it will take a bit of work but I ca...   
29  0010833a96e1f886  "== A barnstar for you! ==\n\n  The Real Life ...   
30  0011cc71398479c4  How could I post before the block expires?  Th...   
31  00128363e367d703  Not sure about a heading of 'Fight for Freedom...   
32  001325b8b20ea8aa  Praise \n\nlooked at this article about 6 mont...   
33  001363e1dbe91225  I was able to post the above list so quickly b...   
34  0013a8b1a5f26bcb  "\nWell, not ""before the process"" but ""befo...   
35  00148d055a169b93  "\n\nNot at all, you are making a straw man ar...   
36  00151a9f93c6b059  "\n\n ""Mainland Asia"" includes ""the lower b...   
37  0015f4aa35ebe9b5  pretty much everyone from warren county/surrou...   
38  00169857adbc989b  Hi Explicit, can you block O Fenian for edit-w...   
39  0016e01b742b8da3  Notability of Rurika Kasuga\nA tag has been pl...   
40  001735f961a23fc4  "\n Sure, but the lead must briefly summarize ...   
41  00173958f46763a2  TFD \n\nI think we just eced. I think we respo...   
42  001810bf8c45bf5f  You are gay or antisemmitian? \n\nArchangel WH...   
43  00190820581d90ce           FUCK YOUR FILTHY MOTHER IN THE ASS, DRY!   
44  001956c382006abd  I'm Sorry \n\nI'm sorry I screwed around with ...   
45  001b2dd65d9d925c  I don't believe the Lisak criticism present th...   
46  001c419c445b5a59  You had a point, and it's now ammended with ap...   
47  001c557175094f10  In other words, you're too lazy to actually po...   
48  001cadfd324f8087  "\nAs for your claims of ""stalking"", that is...   
49  001d874a4d3e8813  "::::Jmabel; in regards to predominant scholar...   

    toxic  severe_toxic  obscene  threat  insult  identity_hate  
0       0             0        0       0       0              0  
1       0             0        0       0       0            

In [ ]:
# get shape of data
df.shape 

(159571, 8)

#### There is 159571 rows and 8 columns

### Information of each column

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   id             159571 non-null  object
 1   comment_text   159571 non-null  object
 2   toxic          159571 non-null  int64 
 3   severe_toxic   159571 non-null  int64 
 4   obscene        159571 non-null  int64 
 5   threat         159571 non-null  int64 
 6   insult         159571 non-null  int64 
 7   identity_hate  159571 non-null  int64 
dtypes: int64(6), object(2)
memory usage: 9.7+ MB


### checking for nulls

In [ ]:
# get the number of nulls in each column
df.isna().sum()

id               0
comment_text     0
toxic            0
severe_toxic     0
obscene          0
threat           0
insult           0
identity_hate    0
dtype: int64

#### there is no nulls in the data

### Describtion of data

In [ ]:
# get the describtion of statistics of the int columns
df.describe().T

count      mean       std  min  25%  50%  75%  max
toxic          159571.0  0.095844  0.294379  0.0  0.0  0.0  0.0  1.0
severe_toxic   159571.0  0.009996  0.099477  0.0  0.0  0.0  0.0  1.0
obscene        159571.0  0.052948  0.223931  0.0  0.0  0.0  0.0  1.0
threat         159571.0  0.002996  0.054650  0.0  0.0  0.0  0.0  1.0
insult         159571.0  0.049364  0.216627  0.0  0.0  0.0  0.0  1.0
identity_hate  159571.0  0.008805  0.093420  0.0  0.0  0.0  0.0  1.0

### Types of columns

In [ ]:
# get types of each column
df.dtypes

id               object
comment_text     object
toxic             int64
severe_toxic      int64
obscene           int64
threat            int64
insult            int64
identity_hate     int64
dtype: object

#### All columns are int except ID and comment_text

### Drop Duplicates (if there is duplicates)

In [ ]:
# drop identical rows
df.drop_duplicates(subset='comment_text',inplace=True) 

In [ ]:
df.shape 

(159571, 8)

### See unique values in data

In [ ]:
# get the number of unique value of each column
df.nunique()

id               159571
comment_text     159571
toxic                 2
severe_toxic          2
obscene               2
threat                2
insult                2
identity_hate         2
dtype: int64

### see the number of unique values in each int column

In [ ]:
# see the quantity of each unique value
df['toxic'].value_counts()

toxic
0    144277
1     15294
Name: count, dtype: int64

In [ ]:
# see the quantity of each unique value
df['severe_toxic'].value_counts()

severe_toxic
0    157976
1      1595
Name: count, dtype: int64

In [ ]:
# see the quantity of each unique value
df['obscene'].value_counts()

obscene
0    151122
1      8449
Name: count, dtype: int64

In [ ]:
# see the quantity of each unique value
df['threat'].value_counts()

threat
0    159093
1       478
Name: count, dtype: int64

In [ ]:
# see the quantity of each unique value
df['insult'].value_counts()

insult
0    151694
1      7877
Name: count, dtype: int64

In [ ]:
# see the quantity of each unique value
df['identity_hate'].value_counts()

identity_hate
0    158166
1      1405
Name: count, dtype: int64

#### I can say from the above cells that all rows has unique id and comment_text , and the rest columns have only zero or one

###  see the rows that have inappropriate comments

In [ ]:
# get the rows that is appropriate by making a condition that all int values = 0
condition = (df["toxic"]==0)&(df["severe_toxic"]==0)&(df["obscene"]==0)&(df["threat"]==0)&(df["insult"]==0)&(df["identity_hate"]==0)

df[condition]

id                                       comment_text  \
0       0000997932d777bf  Explanation\nWhy the edits made under my usern...   
1       000103f0d9cfb60f  D'aww! He matches this background colour I'm s...   
2       000113f07ec002fd  Hey man, I'm really not trying to edit war. It...   
3       0001b41b1c6bb37e  "\nMore\nI can't make any real suggestions on ...   
4       0001d958c54c6e35  You, sir, are my hero. Any chance you remember...   
...                  ...                                                ...   
159566  ffe987279560d7ff  ":::::And for the second time of asking, when ...   
159567  ffea4adeee384e90  You should be ashamed of yourself \n\nThat is ...   
159568  ffee36eab5c267c9  Spitzer \n\nUmm, theres no actual article for ...   
159569  fff125370e4aaaf3  And it looks like it was actually you who put ...   
159570  fff46fc426af1f9a  "\nAnd ... I really don't think you understand...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
0           0             0        0       0       0              0  
1           0             0        0       0       0              0  
2           0             0        0       0       0              0  
3           0             0        0       0       0              0  
4           0             0        0       0       0              0  
...       ...           ...      ...     ...     ...            ...  
159566      0             0        0       0       0              0  
159567      0             0        0       0       0              0  
159568      0             0        0       0       0              0  
159569      0             0        0       0       0              0  
159570      0             0        0       0       0              0  

[143346 rows x 8 columns]

In [ ]:
# get the number of rows that meets the condition
num = df[condition].shape[0] 

# divide the number by whole length of data * 100 to get percentage
Percentage_of_clean_data = (num / len(df))*100 

# get the percentage of appropriate data
Percentage_of_toxic_data = 100 - Percentage_of_clean_data

print("Percentage of clean data {} % \nPercentage of toxic data {} %".format(Percentage_of_clean_data,Percentage_of_toxic_data))

Percentage of clean data 89.83211235124176 % 
Percentage of toxic data 10.167887648758239 %


## Cleaning The Data

### steps of cleaning the data :
- lowercase the data : to ensure consistency
- Removing URLs
- Handling Contractions , for example change "can't" to "cannot"
- Removing Special Characters 
- Remove punctuation
- Remove Numbers
- stop words : to reduce the input's dimensionality

### Implementation of a function that counts the rows that have  ( URLs , Special Characters , Punctuation , Stop Words , Numbers )

In [ ]:
def count_rows(column):
    # Initialize count = zero
    urls_count = 0
    special_chars_count = 0
    punctuation_count = 0
    stop_word_count = 0
    numeric_count = 0
    
    # make a set of all stop words
    stop_words = set(stopwords.words('english'))

    for text in column:
        # Count URLs
        # if URLs found
        if re.search(r'http\S+|www\S+|https\S+', text):
        # add by one
            urls_count += 1

        # Count special characters except english letters and spaces
        # if special characters found 
        if re.search(r'[^A-Za-z0-9\s]+', text):
        # add by one
            special_chars_count += 1
        
        # Count numbers
        # if number found 
        if  re.search(r'\d', text):
        # add by one
            numeric_count +=1

        # Count punctuation
        # if punctuation found 
        if any(char in string.punctuation for char in text):
        # add by one
            punctuation_count += 1

        # Count stop words
        words = text.split()
        # if the word is stop word found 
        if any(word.lower() in stop_words for word in words):
        # add by one
            stop_word_count += 1

    return urls_count, special_chars_count , punctuation_count , stop_word_count , numeric_count

### Implementation of a function that lowercase the data , Remove URLs , Handling Contractions , Remove Special Characters , Remove punctuation , Remove Numbers

In [ ]:
def clean_data(text):

    # Lowercase the text
    text = text.lower()

    # Remove URLs
    text = re.sub(r'http\S+|www\S+|https\S+', '', text, flags=re.MULTILINE)

    # Handle Contractions
    text = contractions.fix(text)
    
    # Remove punctuation
    text = text.translate(str.maketrans('', '', string.punctuation))
    
    # Remove any special characters except english letters and spaces 
    text=re.sub(r'[^A-Za-z\s]+',' ',text)
    
    # Remove stop words
    # Tokenize the text
    tokens = nltk.word_tokenize(text)
    # create a set of stop words using NLTK libarary
    stop_words = set(nltk.corpus.stopwords.words('english'))
    # iterate over the data and remove any stop word in it
    filtered_tokens = [word for word in tokens if word not in stop_words]
    # Join the filtered tokens back into a string
    text = ' '.join(filtered_tokens)
    
    return text

### count the number of rows that have URLs , Special Characters , punctuation , numbers , and stop words BEFORE clean data function 

In [ ]:
# apply the counting count_rows_urls_and_special_chars_and_punctuation_and_stop_words_and_numbers functions
urls_count, special_chars_count , punctuation_count , stop_word_count , numeric_count= count_rows(df['comment_text'])

print(f"Number of rows with URLs: {urls_count}")
print(f"Number of rows with special characters: {special_chars_count}")
print(f"Number of rows with punctuation: {punctuation_count}")
print(f"Number of rows with stop words: {stop_word_count}")
print(f"Number of rows with Numbers: {numeric_count}")

Number of rows with URLs: 5151
Number of rows with special characters: 155146
Number of rows with punctuation: 155106
Number of rows with stop words: 154877
Number of rows with Numbers: 51212


### apply clean data function on data

In [ ]:
# apply clean_data function on comment_text column
df['comment_text'] = df['comment_text'].apply(clean_data)

### count the number of rows that have URLs , Special Characters , punctuation , numbers , and stop words AFTER clean data function 

In [ ]:
# apply the counting count_rows_urls_and_special_chars_and_punctuation_and_stop_words_and_numbers functions
urls_count, special_chars_count , punctuation_count , stop_word_count , numeric_count= count_rows(df['comment_text'])

print(f"Number of rows with URLs: {urls_count}")
print(f"Number of rows with special characters: {special_chars_count}")
print(f"Number of rows with punctuation: {punctuation_count}")
print(f"Number of rows with stop words: {stop_word_count}")
print(f"Number of rows with Numbers: {numeric_count}")

Number of rows with URLs: 0
Number of rows with special characters: 0
Number of rows with punctuation: 0
Number of rows with stop words: 0
Number of rows with Numbers: 0


In [ ]:
df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  daww matches background colour seemingly stuck...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
159566  ffe987279560d7ff  second time asking view completely contradicts...   
159567  ffea4adeee384e90               ashamed horrible thing put talk page   
159568  ffee36eab5c267c9  spitzer umm actual article prostitution ring c...   
159569  fff125370e4aaaf3  looks like actually put speedy first version d...   
159570  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
0           0             0        0       0       0              0  
1           0             0        0       0       0              0  
2           0             0        0       0       0              0  
3           0             0        0       0       0              0  
4           0             0        0       0       0              0  
...       ...           ...      ...     ...     ...            ...  
159566      0             0        0       0       0              0  
159567      0             0        0       0       0              0  
159568      0             0        0       0       0              0  
159569      0             0        0       0       0              0  
159570      0             0        0       0       0              0  

[159571 rows x 8 columns]

#### cleaning function will make some rows hold the same text , so will drop duplicantes on column comment text

In [ ]:
# drop duplicates on comment text column
df.drop_duplicates(subset='comment_text',inplace=True) 

# reset index and drop column index that will be produced from reset index
df = df.reset_index(drop=True)

df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  daww matches background colour seemingly stuck...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
157664  ffe987279560d7ff  second time asking view completely contradicts...   
157665  ffea4adeee384e90               ashamed horrible thing put talk page   
157666  ffee36eab5c267c9  spitzer umm actual article prostitution ring c...   
157667  fff125370e4aaaf3  looks like actually put speedy first version d...   
157668  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
0           0             0        0       0       0              0  
1           0             0        0       0       0              0  
2           0             0        0       0       0              0  
3           0             0        0       0       0              0  
4           0             0        0       0       0              0  
...       ...           ...      ...     ...     ...            ...  
157664      0             0        0       0       0              0  
157665      0             0        0       0       0              0  
157666      0             0        0       0       0              0  
157667      0             0        0       0       0              0  
157668      0             0        0       0       0              0  

[157669 rows x 8 columns]

## Toknization

In [ ]:
# function that tokenize the data which is built in 
def tokenize_text(text):
    tokens = word_tokenize(text)
    return tokens

In [ ]:
# apply function on mentioned column 
df['tokenized_column'] = df['comment_text'].apply(tokenize_text)

In [ ]:
df.head(10)

id                                       comment_text  toxic  \
0  0000997932d777bf  explanation edits made username hardcore metal...      0   
1  000103f0d9cfb60f  daww matches background colour seemingly stuck...      0   
2  000113f07ec002fd  hey man really trying edit war guy constantly ...      0   
3  0001b41b1c6bb37e  make real suggestions improvement wondered sec...      0   
4  0001d958c54c6e35                      sir hero chance remember page      0   
5  00025465d4725e87           congratulations well use tools well talk      0   
6  0002bcb3da6cb337                        cocksucker piss around work      1   
7  00031b1e95af7921  vandalism matt shirvington article reverted pl...      0   
8  00037261f536c51d  sorry word nonsense offensive anyway intending...      0   
9  00040093b2687caa               alignment subject contrary dulithgow      0   

   severe_toxic  obscene  threat  insult  identity_hate  \
0             0        0       0       0              0   
1             0        0       0       0              0   
2             0        0       0       0              0   
3             0        0       0       0              0   
4             0        0       0       0              0   
5             0        0       0       0              0   
6             1        1       0       1              0   
7             0        0       0       0              0   
8             0        0       0       0              0   
9             0        0       0       0              0   

                                    tokenized_column  
0  [explanation, edits, made, username, hardcore,...  
1  [daww, matches, background, colour, seemingly,...  
2  [hey, man, really, trying, edit, war, guy, con...  
3  [make, real, suggestions, improvement, wondere...  
4                [sir, hero, chance, remember, page]  
5    [congratulations, well, use, tools, well, talk]  
6                   [cocksucker, piss, around, work]  
7  [vandalism, matt, shirvington, article, revert...  
8  [sorry, word, nonsense, offensive, anyway, int...  
9          [alignment, subject, contrary, dulithgow]

## Lemmatization 
### To return the word to it's basic form

In [ ]:
def lemmatize_tokens(tokens):
    # initialize WordNetLemmatizer
    lemmatizer = WordNetLemmatizer()

    # Lemmatize each word in the list of tokens
    lemmatized_tokens = [lemmatizer.lemmatize(word) for word in tokens]

    # make the words into sentence instead of a list
    return ' '.join(lemmatized_tokens)

In [ ]:
# apply function on tokenized column
df['lemmatization_column']=df['tokenized_column'].apply(lemmatize_tokens)

In [ ]:
df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  daww matches background colour seemingly stuck...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
157664  ffe987279560d7ff  second time asking view completely contradicts...   
157665  ffea4adeee384e90               ashamed horrible thing put talk page   
157666  ffee36eab5c267c9  spitzer umm actual article prostitution ring c...   
157667  fff125370e4aaaf3  looks like actually put speedy first version d...   
157668  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  \
0           0             0        0       0       0              0   
1           0             0        0       0       0              0   
2           0             0        0       0       0              0   
3           0             0        0       0       0              0   
4           0             0        0       0       0              0   
...       ...           ...      ...     ...     ...            ...   
157664      0             0        0       0       0              0   
157665      0             0        0       0       0              0   
157666      0             0        0       0       0              0   
157667      0             0        0       0       0              0   
157668      0             0        0       0       0              0   

                                         tokenized_column  \
0       [explanation, edits, made, username, hardcore,...   
1       [daww, matches, background, colour, seemingly,...   
2       [hey, man, really, trying, edit, war, guy, con...   
3       [make, real, suggestions, improvement, wondere...   
4                     [sir, hero, chance, remember, page]   
...                                                   ...   
157664  [second, time, asking, view, completely, contr...   
157665        [ashamed, horrible, thing, put, talk, page]   
157666  [spitzer, umm, actual, article, prostitution, ...   
157667  [looks, like, actually, put, speedy, first, ve...   
157668  [really, think, understand, came, idea, bad, r...   

                                     lemmatization_column  
0       explanation edits made username hardcore metal...  
1       daww match background colour seemingly stuck t...  
2       hey man really trying edit war guy constantly ...  
3       make real suggestion improvement wondered sect...  
4                           sir hero chance remember page  
...                                                   ...  
157664  second time asking view completely contradicts...  
157665               ashamed horrible thing put talk page  
157666  spitzer umm actual article prostitution ring c...  
157667  look like actually put speedy first version de...  
157668  really think understand came idea bad right aw...  

[157669 rows x 10 columns]

#### Lemmatization function will make some rows hold the same text , so will drop duplicantes on column lemmatization

In [ ]:
# drop duplicates on lemmatization column
df.drop_duplicates(subset='lemmatization_column',inplace=True) 

# reset index and drop column index that will be produced from reset index
df = df.reset_index(drop=True)

df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  daww matches background colour seemingly stuck...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
157627  ffe987279560d7ff  second time asking view completely contradicts...   
157628  ffea4adeee384e90               ashamed horrible thing put talk page   
157629  ffee36eab5c267c9  spitzer umm actual article prostitution ring c...   
157630  fff125370e4aaaf3  looks like actually put speedy first version d...   
157631  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  \
0           0             0        0       0       0              0   
1           0             0        0       0       0              0   
2           0             0        0       0       0              0   
3           0             0        0       0       0              0   
4           0             0        0       0       0              0   
...       ...           ...      ...     ...     ...            ...   
157627      0             0        0       0       0              0   
157628      0             0        0       0       0              0   
157629      0             0        0       0       0              0   
157630      0             0        0       0       0              0   
157631      0             0        0       0       0              0   

                                         tokenized_column  \
0       [explanation, edits, made, username, hardcore,...   
1       [daww, matches, background, colour, seemingly,...   
2       [hey, man, really, trying, edit, war, guy, con...   
3       [make, real, suggestions, improvement, wondere...   
4                     [sir, hero, chance, remember, page]   
...                                                   ...   
157627  [second, time, asking, view, completely, contr...   
157628        [ashamed, horrible, thing, put, talk, page]   
157629  [spitzer, umm, actual, article, prostitution, ...   
157630  [looks, like, actually, put, speedy, first, ve...   
157631  [really, think, understand, came, idea, bad, r...   

                                     lemmatization_column  
0       explanation edits made username hardcore metal...  
1       daww match background colour seemingly stuck t...  
2       hey man really trying edit war guy constantly ...  
3       make real suggestion improvement wondered sect...  
4                           sir hero chance remember page  
...                                                   ...  
157627  second time asking view completely contradicts...  
157628               ashamed horrible thing put talk page  
157629  spitzer umm actual article prostitution ring c...  
157630  look like actually put speedy first version de...  
157631  really think understand came idea bad right aw...  

[157632 rows x 10 columns]

## BERT

### new column is made containing binary if it has offensive comment or not

#### made offensive column because it is more practical to have binary classification model that tells you this is offensive or non-offensive comment for real life applications 

In [ ]:
# Create a new column with make value equal 1
df['offensive'] = 1

# if it meet the condition and doesnot has offensive comment make the value equal 0
df.loc[condition, 'offensive'] = 0

df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  daww matches background colour seemingly stuck...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
157627  ffe987279560d7ff  second time asking view completely contradicts...   
157628  ffea4adeee384e90               ashamed horrible thing put talk page   
157629  ffee36eab5c267c9  spitzer umm actual article prostitution ring c...   
157630  fff125370e4aaaf3  looks like actually put speedy first version d...   
157631  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  \
0           0             0        0       0       0              0   
1           0             0        0       0       0              0   
2           0             0        0       0       0              0   
3           0             0        0       0       0              0   
4           0             0        0       0       0              0   
...       ...           ...      ...     ...     ...            ...   
157627      0             0        0       0       0              0   
157628      0             0        0       0       0              0   
157629      0             0        0       0       0              0   
157630      0             0        0       0       0              0   
157631      0             0        0       0       0              0   

                                         tokenized_column  \
0       [explanation, edits, made, username, hardcore,...   
1       [daww, matches, background, colour, seemingly,...   
2       [hey, man, really, trying, edit, war, guy, con...   
3       [make, real, suggestions, improvement, wondere...   
4                     [sir, hero, chance, remember, page]   
...                                                   ...   
157627  [second, time, asking, view, completely, contr...   
157628        [ashamed, horrible, thing, put, talk, page]   
157629  [spitzer, umm, actual, article, prostitution, ...   
157630  [looks, like, actually, put, speedy, first, ve...   
157631  [really, think, understand, came, idea, bad, r...   

                                     lemmatization_column  offensive  
0       explanation edits made username hardcore metal...          0  
1       daww match background colour seemingly stuck t...          0  
2       hey man really trying edit war guy constantly ...          0  
3       make real suggestion improvement wondered sect...          0  
4                           sir hero chance remember page          0  
...                                                   ...        ...  
157627  second time asking view completely contradicts...          0  
157628               ashamed horrible thing put talk page          0  
157629  spitzer umm actual article prostitution ring c...          0  
157630  look like actually put speedy first version de...          0  
157631  really think understand came idea bad right aw...          1  

[157632 rows x 11 columns]

### New dataframe is made to enter BERT model containing two columns the comments and offensive column

In [ ]:
# make new dataframe containing only needed columns for bert model
new_df = df.loc[:, ['lemmatization_column','offensive']]

new_df

lemmatization_column  offensive
0       explanation edits made username hardcore metal...          0
1       daww match background colour seemingly stuck t...          0
2       hey man really trying edit war guy constantly ...          0
3       make real suggestion improvement wondered sect...          0
4                           sir hero chance remember page          0
...                                                   ...        ...
157627  second time asking view completely contradicts...          0
157628               ashamed horrible thing put talk page          0
157629  spitzer umm actual article prostitution ring c...          0
157630  look like actually put speedy first version de...          0
157631  really think understand came idea bad right aw...          1

[157632 rows x 2 columns]

### keras 

In [ ]:
# Split the data into input text and labels of the column offensive and get value as numpy array
text = new_df['lemmatization_column'].values
labels = new_df['offensive'].values

# Convert labels to binary matrix
labels = to_categorical(labels)


In [ ]:
# Split the data train and test 
train_text, test_text, train_labels, test_labels = train_test_split(text, labels, test_size=0.2, random_state=42)

In [ ]:
# download pretrained model for BERT
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
bert_model = TFBertModel.from_pretrained('bert-base-uncased')

Some weights of the PyTorch model were not used when initializing the TF 2.0 model TFBertModel: ['cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.dense.bias', 'cls.seq_relationship.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.bias', 'cls.predictions.transform.dense.weight']
- This IS expected if you are initializing TFBertModel from a PyTorch model trained on another task or with another architecture (e.g. initializing a TFBertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TFBertModel from a PyTorch model that you expect to be exactly identical (e.g. initializing a TFBertForSequenceClassification model from a BertForSequenceClassification model).
All the weights of TFBertModel were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFBertModel for predictions w

In [ ]:
# Maximum sequence length for BERT
max_length = 128  

# tokenize the text train and test data
train_encodings = tokenizer.batch_encode_plus(train_text.tolist(), max_length=max_length, padding=True, truncation=True)
test_encodings = tokenizer.batch_encode_plus(test_text.tolist(), max_length=max_length, padding=True, truncation=True)

# np.array(encodings['input_ids']) NumPy array contains the tokenized input sequences converted into integer
# np.array(train_encodings['attention_mask']) NumPy array contains the attention mask 
train_inputs = [np.array(train_encodings['input_ids']), np.array(train_encodings['attention_mask'])]
test_inputs = [np.array(test_encodings['input_ids']), np.array(test_encodings['attention_mask'])]

In [ ]:
# input_ids and attention_mask are the input layers in the neural networks and input length is 128
input_ids = Input(shape=(max_length,), dtype=tf.int32, name="input_ids")
attention_mask = Input(shape=(max_length,), dtype=tf.int32, name="attention_mask")

# put the input layers inn the pretrained model
outputs = bert_model(input_ids, attention_mask=attention_mask)

# extracts the pooled output from the BERT model output
pooled_output = outputs[1]

# output layer is 2 nodes with softmax as an activation function
output = Dense(2, activation='softmax')(pooled_output)  

# define model input and output
model = Model(inputs=[input_ids, attention_mask], outputs=output)

In [ ]:
# compile using ADAM optimizer and loss function is cross entropy
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# fit 1 epoch and batch size equal 16
model.fit(train_inputs, train_labels, epochs=1, batch_size=16)  

7882/7882 [==============================] - 49644s 6s/step - loss: 0.3565 - accuracy: 0.8976


In [ ]:
# predict using test data
predictions = model.predict(test_inputs)

# get the higger probabilty to assign the sentence to it from predicted data
predicted_labels = np.argmax(predictions, axis=1)

# get the higger probabilty to assign the sentence to it from actual data
true_labels = np.argmax(test_labels, axis=1)

986/986 [==============================] - 3765s 4s/step


In [ ]:
from sklearn.metrics import accuracy_score

# Calculate accuracy
accuracy = accuracy_score(true_labels, predicted_labels)

print(f"Accuracy: {accuracy}")

Accuracy: 0.8992926697751134


#### explaination of the output
- Accuracy = 90 % which is good